application/templates/_helpers.tpl

```diff
@@ -39,7 +39,9 @@ Common labels
 */}}
 {{- define "application.labels" -}}
 helm.sh/chart: {{ include "application.chart" . }}
+{{- if .Values.deployment.image.tag }}
 app.kubernetes.io/version: {{ include "application.version" . }}
+{{- end }}
```

Comment:
With an empty string as the default value for `.Values.deployment.image.tag`, we could discard this `if` statement, though we would end up with an empty value for a version label. How about leaving out this conditional and instead editing the `application.version` template so that it:

- Tries to use the tag as it currently does, but failing that, it
- Tries to use the digest with a little bit of RegEx substitution magic, but if not provided either, it
- Simply says "latest", as is default behavior for container images?

How does that sound?

Reply:
> The application version should always be a semantic version.

I'm not sure if I agree with this one. Helm does require chart versions to be semantic, but the versioning scheme for an application is an incredibly arbitrary choice, usually left to the core maintainers or company behind the project. Hence Helm will let `appVersion` be just about anything in `Chart.yaml` — in fact `.Chart.AppVersion` is often used as the default value when an image tag isn't provided, but this is a generic chart and such a value can't be set during installation.

So that said, whatever logic is present in `application.version`, now or in the future, this could be changed to a simple:

```suggestion
{{- with include "application.version" .  }}
app.kubernetes.io/version: {{ quote . }}
{{- end }}
```

What do you think?

Reply:
Okay, i see. But what if there is no tag defined? Maybe the tag is a good default for the label, but we could allow overriding the label - maybe with another Value, e.g. `.Values.applicationVersion`? 
Also there could be situations where the version of the application stays the same (e.g. the wordpress version) but the underlying base container receives an update (e.g. packages updates of the alpine/ubi container).

Reply:
An `appVersion` value sounds good to me. In the case of no such value and no tag being defined, however, we can just skip the label altogether as it makes no sense to set it, and hence the suggested `with` being appropriate.